lxd/storage/drivers/driver_powerflex_utils.go

```diff
@@ -0,0 +1,1021 @@
+func (p *powerFlexClient) getStoragePool(poolID string) (*powerFlexStoragePool, error) {
+	var actualResponse powerFlexStoragePool
+	err := p.requestAuthenticated(http.MethodGet, fmt.Sprintf("/api/instances/StoragePool::%s", poolID), nil, &actualResponse)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to get storage pool: %q: %w", poolID, err)
+	}
+
+	return &actualResponse, nil
+}
+
+// getProtectionDomain returns the protection domain behind domainID.
+func (p *powerFlexClient) getProtectionDomain(domainID string) (*powerFlexProtectionDomain, error) {
+	var actualResponse powerFlexProtectionDomain
+	err := p.requestAuthenticated(http.MethodGet, fmt.Sprintf("/api/instances/ProtectionDomain::%s", domainID), nil, &actualResponse)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to get protection domain: %q: %w", domainID, err)
+	}
+
+	return &actualResponse, nil
+}
+
+// getProtectionDomainSdsRelations returns the protection domains SDS relations.
+func (p *powerFlexClient) getProtectionDomainSdsRelations(domainID string) ([]powerFlexProtectionDomainSdsRelation, error) {
+	var actualResponse []powerFlexProtectionDomainSdsRelation
+	err := p.requestAuthenticated(http.MethodGet, fmt.Sprintf("/api/instances/ProtectionDomain::%s/relationships/Sds", domainID), nil, &actualResponse)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to get protection domain SDS relations: %q: %w", domainID, err)
+	}
+
+	return actualResponse, nil
+}
+
+// getVolumeID returns the volume ID for the given name.
+func (p *powerFlexClient) getVolumeID(name string) (string, error) {
+	body, err := p.createBodyReader(map[string]any{
+		"name": name,
+	})
+	if err != nil {
+		return "", err
+	}
+
+	var actualResponse string
+	err = p.requestAuthenticated(http.MethodPost, "/api/types/Volume/instances/action/queryIdByKey", body, &actualResponse)
+	if err != nil {
+		powerFlexError, ok := err.(*powerFlexError)
+		if ok {
+			// API returns 500 if the volume does not exist.
+			// To not confuse it with other 500 that might occur check the error code too.
+			if powerFlexError.HTTPStatusCode() == 500 && powerFlexError.ErrorCode() == 79 {
+				return "", api.StatusErrorf(http.StatusNotFound, "PowerFlex volume not found")
+			}
+		}
+
+		return "", fmt.Errorf("Failed to get volume ID: %q: %w", name, err)
+	}
+
+	return actualResponse, nil
+}
+
+// getVolume returns the volume behind volumeID.
+func (p *powerFlexClient) getVolume(volumeID string) (*powerFlexVolume, error) {
+	var actualResponse powerFlexVolume
+	err := p.requestAuthenticated(http.MethodGet, fmt.Sprintf("/api/instances/Volume::%s", volumeID), nil, &actualResponse)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to get volume: %q: %w", volumeID, err)
+	}
+
+	return &actualResponse, nil
+}
+
+// createVolume creates a new volume.
+// The returned string represents the ID of the volume.
+func (p *powerFlexClient) createVolume(name string, size string, volumeType string, poolID string) (string, error) {
```

Comment:
Can we make `volumeType` either a bool and call it `thin bool` or something like that, or introduce it as its own int type called `powerflexVolumeType` or similar with constants, as we already have the concept of "volume type" in LXD and when I looked at this signature I initially thought it meant that and only looking at the implementation did I realise its a totally different type

Reply:
For example LVM driver has `createLogicalVolume(vgName, thinPoolName string, vol Volume, makeThinLv bool)`

Reply:
That is confusing indeed. Haven't thought about the meaning in LXD. I'll go with some constants that will represent the actual string required by the PowerFlex API. Having a boolean feels biased. Which type would be `true` then.

Reply:
Fixed. Both the volume type (thin, thick) and access mode (rw) are now covered by constants.